### PR TITLE
Using sea level atmospheric pressure for HPWH ratings

### LINKED ISSUE #5110: HPWH Rated Coil Bypass Factor being calculated at simulation elevation instead of sea level
When a heat pump water heater is used at anywhere at elevation, the `StdBaroPress` for that elevation is used to calculate the rated coil bypass factor for that unit when it should use the barometric pressure at sea level. 

```
   ** Severe  ** For object = Coil:WaterHeating:AirToWaterHeatPump:Wrapped, name = "HPWH COIL_1"
   **   ~~~   ** Calculated outlet air relative humidity greater than 1. The combination of
   **   ~~~   ** rated air volume flow rate, total cooling capacity and sensible heat ratio yields coil exiting
   **   ~~~   ** air conditions above the saturation curve. Possible fixes are to reduce the rated total cooling
   **   ~~~   ** capacity, increase the rated air volume flow rate, or reduce the rated sensible heat ratio for this coil.
   **   ~~~   ** If autosizing, it is recommended that all three of these values be autosized.
   **   ~~~   ** ...Inputs used for calculating cooling coil bypass factor.
   **   ~~~   ** ...Inlet Air Temperature     = 19.72 C
   **   ~~~   ** ...Outlet Air Temperature    = 6.20 C
   **   ~~~   ** ...Inlet Air Humidity Ratio  = 8.756020E-003 kgWater/kgDryAir
   **   ~~~   ** ...Outlet Air Humidity Ratio = 8.015009E-003 kgWater/kgDryAir
   **   ~~~   ** ...Total Cooling Capacity used in calculation = 1303.60 W
   **   ~~~   ** ...Air Mass Flow Rate used in calculation     = 8.319111E-002 kg/s
   **   ~~~   ** ...Air Volume Flow Rate used in calculation   = 8.542249E-002 m3/s
   **   ~~~   ** ...Air Volume Flow Rate per Watt of Rated Cooling Capacity is also out of bounds at = 6.5528207E-005 m3/s/W
   **   ~~~   **  During Warmup, Environment=PLANT DD, at Simulation time=01/01 00:00 - 00:01
```

@jmaguire1

### PR DESCRIPTION
**Pull request by nmerket**
Fixes #5110

### CODE REVIEW

File: src/EnergyPlus/DXCoils.cc [hunk 1]
```diff
@@ -5211,10 +5212,10 @@ namespace DXCoils {
...
 				ShowContinueError( "Min Rated Vol Flow Per Watt=[" + TrimSigDigits( MinRatedVolFlowPerRatedTotCap( DXCT ), 3 ) + "], Rated Vol Flow Per Watt=[" + TrimSigDigits( RatedVolFlowPerRatedTotCap, 3 ) + "], Max Rated Vol Flow Per Watt=[" + TrimSigDigits( MaxHeatVolFlowPerRatedTotCap( DXCT ), 3 ) + "]. See Input-Output Reference Manual for valid range." );
 			}
-			HPInletAirHumRat = PsyWFnTdbTwbPb( DXCoil( DXCoilNum ).RatedInletDBTemp, DXCoil( DXCoilNum ).RatedInletWBTemp, StdBaroPress, RoutineName );
+			HPInletAirHumRat = PsyWFnTdbTwbPb( DXCoil( DXCoilNum ).RatedInletDBTemp, DXCoil( DXCoilNum ).RatedInletWBTemp, StdBaroPressSeaLevel, RoutineName );
 			HPWHInletDBTemp = DXCoil( DXCoilNum ).RatedInletDBTemp;
 			HPWHInletWBTemp = DXCoil( DXCoilNum ).RatedInletWBTemp;
-			DXCoil( DXCoilNum ).RatedAirMassFlowRate( 1 ) = DXCoil( DXCoilNum ).RatedAirVolFlowRate( 1 ) * PsyRhoAirFnPbTdbW( StdBaroPress, DXCoil( DXCoilNum ).RatedInletDBTemp, HPInletAirHumRat, RoutineName );
+			DXCoil( DXCoilNum ).RatedAirMassFlowRate( 1 ) = DXCoil( DXCoilNum ).RatedAirVolFlowRate( 1 ) * PsyRhoAirFnPbTdbW( StdBaroPressSeaLevel, DXCoil( DXCoilNum ).RatedInletDBTemp, HPInletAirHumRat, RoutineName );
```
Comment (mjwitte) [inline src/EnergyPlus/DXCoils.cc]: @nmerket This appears to be inconsistent with the general EnergyPlus rule that all user inputs in volume flow rate get changed to massflow using the standard density at the location elevation.  If I'm following correctly, if a user autosizes the coil or enters the same rated volume flow rate here and in the airloop, they will expect the coil to operate with a flow fraction of 1.0 at full flow.  This change will cause a deviation from that.

Comment (rraustad) [inline src/EnergyPlus/DXCoils.cc]: I wouldn't do this here, I'd do it in the CBF calculation (i.e., back out rho at elevation and use sea level instead).

File: src/EnergyPlus/DataEnvironment.hh [hunk 2]
```diff
@@ -22,6 +22,7 @@ namespace DataEnvironment {
 	extern Real64 const EarthRadius; // Radius of the Earth (m)
 	extern Real64 const AtmosphericTempGradient; // Standard atmospheric air temperature gradient (K/m)
 	extern Real64 const SunIsUpValue; // if Cos Zenith Angle of the sun is >= this value, the sun is "up"
+	extern Real64 const StdBaroPressSeaLevel; // Standard barometric pressure at sea level (Pa)
```
Comment (EnergyArchmage) [inline src/EnergyPlus/DataEnvironment.hh]: I happen to have just drafted out similar changes and came up with different names and to also add to weather manager a one time calculation of rho at STP.

```
extern Real64 const stdAtmosphericPressure; // one atmosphere, sea level standard pressure 101325 Pa
extern Real64 rhoAirSTP; // Standard density of dry air at 101325 Pa, 20.0C temperaure 
```

File: src/EnergyPlus/WeatherManager.cc [hunk 3]
```diff
@@ -4632,8 +4632,7 @@ Label9999: ;
 		}
 
 		if ( ! ErrorsFound ) {
-			StdBaroPress = 101.325 * std::pow( 1.0 - 2.25577e-05 * Elevation, 5.2559 );
-			StdBaroPress *= 1000.0;
+			StdBaroPress = StdBaroPressSeaLevel * std::pow( 1.0 - 2.25577e-05 * Elevation, 5.2559 );
```
Comment (EnergyArchmage) [inline src/EnergyPlus/WeatherManager.cc]: This is wrong, the correlation is expecting kPa, you can't change to Pa and use the same equation.

File: src/EnergyPlus/DataEnvironment.hh (hunk 2 above)
Comment (mjwitte) [inline src/EnergyPlus/DataEnvironment.hh]: @EnergyArchmage I would prefer "sea level" in the name to help clarify.  "std" and "STP" are too ambiguous here.  The comments are clear, but that gets lost when used elsewhere in the code.

File: src/EnergyPlus/DXCoils.cc (hunk 1 above)
Comment (nmerket) [inline src/EnergyPlus/DXCoils.cc]: Okay, I'll look into making that change.

File: src/EnergyPlus/DataEnvironment.hh (hunk 2 above)
Comment (nmerket) [inline src/EnergyPlus/DataEnvironment.hh]: Sea level seems more clear to me (probably why I chose that for my constant name), but I'm fine with whatever as long as we all agree to use the same thing.

File: src/EnergyPlus/WeatherManager.cc (hunk 3 above)
Comment (nmerket) [inline src/EnergyPlus/WeatherManager.cc]: @EnergyArchmage, it is mathematically the same equation. It just does in one line now what was previously done in two. I can change it back, though, if you really want.

File: src/EnergyPlus/DXCoils.cc [hunk 4]
```diff
@@ -5211,7 +5212,7 @@ namespace DXCoils {
 				ShowSevereError( DXCoil( DXCoilNum ).DXCoilType + " \"" + DXCoil( DXCoilNum ).Name + "\": Rated air volume flow rate per watt of rated total water heating capacity is out of range" );
 				ShowContinueError( "Min Rated Vol Flow Per Watt=[" + TrimSigDigits( MinRatedVolFlowPerRatedTotCap( DXCT ), 3 ) + "], Rated Vol Flow Per Watt=[" + TrimSigDigits( RatedVolFlowPerRatedTotCap, 3 ) + "], Max Rated Vol Flow Per Watt=[" + TrimSigDigits( MaxHeatVolFlowPerRatedTotCap( DXCT ), 3 ) + "]. See Input-Output Reference Manual for valid range." );
 			}
-			HPInletAirHumRat = PsyWFnTdbTwbPb( DXCoil( DXCoilNum ).RatedInletDBTemp, DXCoil( DXCoilNum ).RatedInletWBTemp, StdBaroPress, RoutineName );
+			HPInletAirHumRat = PsyWFnTdbTwbPb( DXCoil( DXCoilNum ).RatedInletDBTemp, DXCoil( DXCoilNum ).RatedInletWBTemp, StdBaroPressSeaLevel, RoutineName );
 			HPWHInletDBTemp = DXCoil( DXCoilNum ).RatedInletDBTemp;
 			HPWHInletWBTemp = DXCoil( DXCoilNum ).RatedInletWBTemp;
 			DXCoil( DXCoilNum ).RatedAirMassFlowRate( 1 ) = DXCoil( DXCoilNum ).RatedAirVolFlowRate( 1 ) * PsyRhoAirFnPbTdbW( StdBaroPress, DXCoil( DXCoilNum ).RatedInletDBTemp, HPInletAirHumRat, RoutineName );
```
Comment (nmerket) [inline src/EnergyPlus/DXCoils.cc:5441]: @mjwitte @rraustad Changed back to use air density at elevation to calculate the `RatedAirMassFlowRate`.

File: src/EnergyPlus/DataEnvironment.hh (hunk 2 above)
Comment (EnergyArchmage) [inline src/EnergyPlus/DataEnvironment.hh]: if I cared, which I don't, I'd point out that the word choice of barometric pressure to me is something that varies with the weather and not the same as a standard atmosphere of press.  Like when a hurricane is coming the barometric pressure at sea level drops so it isn't a fixed value of Pa.  But since I don't care I 'll just merge to use whatever you want to call it.

File: src/EnergyPlus/DXCoils.cc [hunk 5]
```diff
@@ -5224,7 +5225,8 @@ namespace DXCoils {
 				MySizeFlag( DXCoilNum ) = false;
 			}
 
-			DXCoil( DXCoilNum ).RatedCBF( 1 ) = CalcCBF( DXCoil( DXCoilNum ).DXCoilType, DXCoil( DXCoilNum ).Name, DXCoil( DXCoilNum ).RatedInletDBTemp, HPInletAirHumRat, DXCoil( DXCoilNum ).RatedTotCap( 1 ), DXCoil( DXCoilNum ).RatedAirMassFlowRate( 1 ), DXCoil( DXCoilNum ).RatedSHR( 1 ) );
+			Real64 const RatedAirMassFlowRateSeaLevel = DXCoil( DXCoilNum ).RatedAirVolFlowRate( 1 ) * PsyRhoAirFnPbTdbW( StdBaroPressSeaLevel, DXCoil( DXCoilNum ).RatedInletDBTemp, HPInletAirHumRat, RoutineName );
+			DXCoil( DXCoilNum ).RatedCBF( 1 ) = CalcCBF( DXCoil( DXCoilNum ).DXCoilType, DXCoil( DXCoilNum ).Name, DXCoil( DXCoilNum ).RatedInletDBTemp, HPInletAirHumRat, DXCoil( DXCoilNum ).RatedTotCap( 1 ), RatedAirMassFlowRateSeaLevel, DXCoil( DXCoilNum ).RatedSHR( 1 ), StdBaroPressSeaLevel );
```
Comment (nmerket) [inline src/EnergyPlus/DXCoils.cc]: @mjwitte @rraustad Using an air mass flow rate based on sea level air density only for the rated CBF calculation now.

File: src/EnergyPlus/DataEnvironment.hh (hunk 2 above)
Comment (mjwitte) [inline src/EnergyPlus/DataEnvironment.hh]: @EnergyArchmage That's a valid point.  So, how about stdPressureSeaLevel  or stdAtmPressureSeaLevel ?

Comment (nmerket) [inline src/EnergyPlus/DataEnvironment.hh]: I like `StdPressureSeaLevel` because it's shorter.

**Comment by nmerket** [comment]
I tried to make this change so that it would not change the operation of other objects. However, I think there's a good case to argue whether `CalcCBF` should ever be called for any coil with anything except standard pressure at sea level.

**Comment by nmerket** [comment]
I know this stirred up some controversy when I first brought the issue up, but not having this breaks our implementation of heat pump water heaters for an upcoming version of BEopt for any location above 3000-ish feet. Therefore, I'm bringing it up again. This pull request solves the problem for heat pump water heater coils and opens the possibility to do the same for other coils, but leaves them unchanged for now. 
@mjwitte @rraustad @EnergyArchmage Can any of you take a look at this and let me know if it's an acceptable solution and if not what we can do it make it so?

**Comment by mjwitte** [comment]
@nmerket I can, but there are other review items in the queue before this.  Probably be next week.  Is that soon enough?

File: src/EnergyPlus/DXCoils.cc [hunk 6]
```diff
@@ -9606,10 +9609,10 @@ Label50: ;
...
-					ShowContinueError( "...Air Volume Flow Rate per Watt of Rated Cooling Capacity is also out of bounds at = " + RoundSigDigits( AirMassFlowRate / PsyRhoAirFnPbTdbW( StdBaroPress, InletAirTemp, InletAirHumRat, RoutineName ) / TotCap, 7 ) + " m3/s/W" );
+				if ( ( ( MinRatedVolFlowPerRatedTotCap( DXCT ) - AirMassFlowRate / PsyRhoAirFnPbTdbW( BaroPress, InletAirTemp, InletAirHumRat, RoutineName ) / TotCap ) > SmallDifferenceTest ) || ( ( AirMassFlowRate / PsyRhoAirFnPbTdbW( BaroPress, InletAirTemp, InletAirHumRat, RoutineName ) / TotCap - MaxRatedVolFlowPerRatedTotCap( DXCT ) ) > SmallDifferenceTest ) ) {
+					ShowContinueError( "...Air Volume Flow Rate per Watt of Rated Cooling Capacity is also out of bounds at = " + RoundSigDigits( AirMassFlowRate / PsyRhoAirFnPbTdbW( BaroPress, InletAirTemp, InletAirHumRat, RoutineName ) / TotCap, 7 ) + " m3/s/W" );
```
Comment (rraustad) [inline src/EnergyPlus/DXCoils.cc:9887]: The calculation of VolFlowPerRatedTotCap needs to use the same pressure used to convert user input volume flow rate to mass flow rate. Otherwise warnings will pop up due to slight differences in Rho.

line 5303, 5334, 5358, 5376, 5407, 5419:
DXCoil( DXCoilNum ).RatedAirMassFlowRate( Mode ) = DXCoil( DXCoilNum ).RatedAirVolFlowRate( Mode ) \* PsyRhoAirFnPbTdbW( StdBaroPress, RatedInletAirTemp, RatedInletAirHumRat, RoutineName );

If you changed these mass flow calculations to use BaroPress, then I would be OK with it (the change in line 9611).

Comment (rraustad) [inline src/EnergyPlus/DXCoils.cc:9887]: Never mind, BaroPress = StdBaroPress

File: src/EnergyPlus/DXCoils.cc [hunk 7]
```diff
@@ -9591,7 +9594,7 @@ Label50: ;
 		OutletAirTemp = PsyTdbFnHW( OutletAirEnthalpy, OutletAirHumRat );
 		//  Eventually inlet air conditions will be used in DX Coil, these lines are commented out and marked with this comment line
 		//  Pressure will have to be pass into this subroutine to fix this one
-		OutletAirRH = PsyRhFnTdbWPb( OutletAirTemp, OutletAirHumRat, StdBaroPress, RoutineName );
+		OutletAirRH = PsyRhFnTdbWPb( OutletAirTemp, OutletAirHumRat, BaroPress, RoutineName );
```
Comment (rraustad) [inline src/EnergyPlus/DXCoils.cc:9869]: Now that I know that BaroPress = StdBaroPress, why did you need the new variable?

Comment (rraustad) [inline src/EnergyPlus/DXCoils.cc:9869]: So these changes (where you changed StdBaroPress to instead use BaroPress) don't change the answer? So then what's the point?

File: src/EnergyPlus/DXCoils.cc [hunk 8]
```diff
@@ -5224,7 +5225,8 @@ namespace DXCoils {
 				MySizeFlag( DXCoilNum ) = false;
 			}
 
-			DXCoil( DXCoilNum ).RatedCBF( 1 ) = CalcCBF( DXCoil( DXCoilNum ).DXCoilType, DXCoil( DXCoilNum ).Name, DXCoil( DXCoilNum ).RatedInletDBTemp, HPInletAirHumRat, DXCoil( DXCoilNum ).RatedTotCap( 1 ), DXCoil( DXCoilNum ).RatedAirMassFlowRate( 1 ), DXCoil( DXCoilNum ).RatedSHR( 1 ) );
+			Real64 const RatedAirMassFlowRateSeaLevel = DXCoil( DXCoilNum ).RatedAirVolFlowRate( 1 ) * PsyRhoAirFnPbTdbW( StdPressureSeaLevel, DXCoil( DXCoilNum ).RatedInletDBTemp, HPInletAirHumRat, RoutineName );
+			DXCoil( DXCoilNum ).RatedCBF( 1 ) = CalcCBF( DXCoil( DXCoilNum ).DXCoilType, DXCoil( DXCoilNum ).Name, DXCoil( DXCoilNum ).RatedInletDBTemp, HPInletAirHumRat, DXCoil( DXCoilNum ).RatedTotCap( 1 ), RatedAirMassFlowRateSeaLevel, DXCoil( DXCoilNum ).RatedSHR( 1 ), StdPressureSeaLevel );
```
Comment (nmerket) [inline src/EnergyPlus/DXCoils.cc:5452]: @rraustad 

First, I'd like to define a couple of the variables that aren't immediately clear:
- `StdBaroPress` - Standard "atmospheric pressure" based on elevation (ASHRAE HOF p6.1) defined [here](https://github.com/NREL/EnergyPlus/blob/8489490953551511afd4898eea8dde65daa7e906/src/EnergyPlus/DataEnvironment.cc#L132) and [calculated here](https://github.com/NREL/EnergyPlus/blob/8489490953551511afd4898eea8dde65daa7e906/src/EnergyPlus/WeatherManager.cc#L4635) based on the elevation of the current simulation.
- `StdPressureSeaLevel` - One atmosphere at sea level. The standard pressure part of "standard temperature and pressure". 101.325 kPa. Constant. 

The problem was that when calculating the _rated_ coil bypass factor for `Coil:WaterHeating:AirToWaterHeatPump:*`, it was using all of the rated conditions except for the atmospheric pressure. It was using `StdBaroPress` instead, which is incorrect. `CalcCBF` had that hard coded in. The point of these changes was to open up `CalcCBF` to accept a different pressure--`StdPressureSeaLevel`--when I wanted to pass that in (on this line) and to default it to `StdBaroPress` at all other times. This changes the calculation of `CalcCBF` for `Coil:WaterHeating:AitToWaterHeatPump:*` objects only and leaves the calculation for all other DX Coils the same as before assuming they are correctly handling calculation of rated conditions.

**Comment by nmerket** [comment]
@mjwitte Next week is soon enough. Thank you.

**Comment by nmerket** [comment]
@mjwitte @rraustad Just wanted to make sure this is still on your radar.

PS @rraustad did you see my [answer to your questions](https://github.com/NREL/EnergyPlus/pull/5116#discussion_r37665031)?

**Comment by rraustad** [comment]
We'll go back and look at other DX coils after this is checked in.

**Comment by nmerket** [comment]
Yeah, I definitely think that a follow up issue with the remaining coils would be a great idea.

**Comment by mjwitte** [comment]
@nmerket We'll get there.  This week the priority is the new features, then we'll work on the defects, etc.  Lots to push through right now.

**Comment by mjwitte** [comment]
@nmerket Merged in develop to bring this branch up to date.  Will let CI run one more round.  Assuming it's all green, then this can go in.  @rraustad @EnergyArchmage agree?

**Comment by rraustad** [comment]
I'm OK with this.

**Comment by mjwitte** [comment]
@nmerket Surprising that this causes diffs in files without any DX coils or heat pump water heaters, e.g. 5Zone_Transformer.  Possibly due to precision in the StdBaroPressure calculation reordering?  Not a big deal, just puzzling.

File: src/EnergyPlus/DataEnvironment.cc [hunk 9]
```diff
@@ -128,7 +129,7 @@ namespace DataEnvironment {
 	Real64 PDIFLW; // Luminous efficacy (lum/W) of sky diffuse solar radiation
 	Real64 SkyClearness; // Sky clearness (see subr. DayltgLuminousEfficacy)
 	Real64 SkyBrightness; // Sky brightness (see subr. DayltgLuminousEfficacy)
-	Real64 StdBaroPress( 101325.0 ); // Standard "atmospheric pressure" based on elevation (ASHRAE HOF p6.1)
+	Real64 StdBaroPress( StdPressureSeaLevel ); // Standard "atmospheric pressure" based on elevation (ASHRAE HOF p6.1)
 	Real64 StdRhoAir; // Standard "rho air" set in WeatherManager - based on StdBaroPress
 	Real64 const stdAtmosphericPressure( 101325.0 ); // one atmosphere, sea level standard pressure 101325 Pa
```
Comment (nmerket) [inline src/EnergyPlus/DataEnvironment.cc]: Just noticed this here. This is from @EnergyArchmage's branch. Anyway that means we now have two constants with the same meaning/value: `StdPressureSeaLevel` and `stdAtmosphericPressure`. We really shouldn't do that. I'm fine with going with `stdAtmosphericPressure` since it's already in `develop`. I'll just need to go in a make the change. So, don't merge this yet.
@mjwitte

Comment (mjwitte) [inline src/EnergyPlus/DataEnvironment.cc]: @nmerket Well, I think I prefer the explicit SeaLevel in the name, even though AtmosphericPressure technically means at sea level.  How many places does this appear in develop?

**Comment by nmerket** [comment]
@mjwitte Yeah, it is always concerning to see those red dots. The HeatPumpWaterHeater isn't surprising. I was intentionally changing that one. The others seem to be only different [on the eio](http://energyplus.s3-website-us-east-1.amazonaws.com/regressions/2015-09/e9b39c22b7c3477fea1f9ab9857b761ac65684ab-9f8fb4c0247cd554d1c46c2dfb2549fe4c864875/5Zone_Transformer/x86_64-MacOS-10.9-clang/eplusout.eio.diff.html) during warmup at the 7th and 8th significant figures. Seems like noise to me.

**Comment by nmerket** [comment]
@mjwitte I like `StdPressureSeaLevel` better too, so I removed the other and they're all using that one constant now. I think this is ready to go after _one more_ round of CI tests. Turning the reigns over to you.

**Comment by mjwitte** [comment]
@Myoldmopar This is probably ok to merge, but I was waiting for at least one debug CI report.  Do you think CI will pick up on this branch now, or do I need to tweak the branch to get its attention?

**Comment by Myoldmopar** [comment]
@mjwitte it should get picked up automatically.  However, now that the ground model work is in develop, later CI runs will show diffs.  Might be good to just pull develop in anyway at this point.  Want me to do it?  Should be automatic merge according to the green button above.

**Comment by mjwitte** [comment]
Sure, go ahead and merge it up

**Comment by Myoldmopar** [comment]
In response to [@nmerket](https://github.com/NREL/EnergyPlus/pull/5116#issuecomment-138976988), I would note that the eio file is compared as a direct text diff.  Any change is a full failure, even down at the 10-11th decimal place.  Whereas the csv output is compared with MathDiff, so small diffs are filtered differently.  Not saying that is what is happening here, just noting the difference.
